modules/diskq/logqueue-disk.c

```diff
@@ -203,4 +225,86 @@ log_queue_disk_init_instance(LogQueueDisk *self, DiskQueueOptions *options, cons
   self->super.type = log_queue_disk_type;
 
   self->qdisk = qdisk_new(options, qdisk_file_id);
+  _set_serializer(self);
+}
+
+static gchar *
+_request_serializer(SerializeArchive **sa, GString *serialized)
+{
+  *sa = serialize_string_archive_new(serialized);
+
+  /* Leave space for the real record_length for later */
+  if (!serialize_write_uint32(*sa, 0))
+    return "cannot write record length";
+
+  return NULL;
+}
+
+static gboolean
+_serialize_msg(LogQueueDisk *self, LogMessage *msg, SerializeArchive **sa)
+{
+  if (self->serialize_msg)
+    return self->serialize_msg(msg, *sa);
+
+  return FALSE;
+}
+
+static gchar *
+_overwrite_with_real_record_length(GString *serialized)
+{
+  guint32 record_length = GUINT32_TO_BE(serialized->len - sizeof(guint32));
+  if (record_length == 0)
+    return "message is empty";
+
+  g_string_overwrite_len(serialized, 0, (gchar *) &record_length, sizeof(guint32));
+  return NULL;
+}
+
+gboolean
+log_queue_disk_serialize_msg(LogQueueDisk *self, LogMessage *msg, GString *serialized)
+{
+  gchar *error = NULL;
+  SerializeArchive *sa;
+
+  if ((error = _request_serializer(&sa, serialized)) != NULL)
+    goto exit;
+
+  if (!_serialize_msg(self, msg, &sa))
+    {
+      error = "cannot serialize LogMessage";
+      goto exit;
+    }
+
+  if ((error = _overwrite_with_real_record_length(serialized)) != NULL)
+    goto exit;
+
+exit:
+  if (error)
+    {
+      msg_error("Error serializing message for the disk-queue file",
+                evt_tag_str("error", error),
+                evt_tag_str("filename", log_queue_disk_get_filename(&self->super)));
```

Comment:
The `log_queue_disk_get_filename` still access qdisk without locking (not asking for lock here), that yields the same segfault as this PR aims to solve.